cs/src/core/Index/FasterLog/FasterLog.cs

```diff
@@ -1166,28 +1163,27 @@ private void UpdateTailCallback(long tailAddress)
         }
 
         
-        // TODO(Tianyu): Will we ever need to recover to a specific commit for read-only cases?
         /// <summary>
-        /// Synchronously recover instance to FasterLog's latest commit, when being used as a readonly log iterator
+        /// Synchronously recover instance to FasterLog's latest valid commit, when being used as a readonly log iterator
         /// </summary>
-        public void RecoverReadOnly()
+        public void RecoverReadOnly(bool purgeEarlierCommits = false)
```

Comment:
i feel like this might be a bit weird to allow readonly clients to purge commits? Definitely smells like a write operation to me. This is especially weird because I don't think the non-readonly FasterLog has a similar option.

Reply:
The real question is, how then can we purge commits. Having the writer purge commits makes it impossible for readers to consistently read the latest as it gets deleted after reader discovers it and before it reads.  Or worse,  exceptions thrown if writer deletes when reader reading.

Reply:
The same problem is present with this version if we have multiple readers right?

The dumb and easy fix in this case seems to be have readers retry on exceptions thrown -- purged commits will always be replaced by a later commit.

Also, I am not certain about this: do read only consumers of FasterLog automatically catch up to newly committed tails, or do they need to continuously invoke RecoverReadOnly()? If latter, this might be a bad idea for fastcommits due to double scanning.

Reply:
> The same problem is present with this version if we have multiple readers right?

correct, that is why it is exposed to the API so that only the last reader would invoke it with purge=true.

> do read only consumers of FasterLog automatically catch up to newly committed tails, or do they need to continuously invoke RecoverReadOnly()?

It is the latter. Each call to RecoverReadOnly moves the reader's view forward as it accesses the latest commit.dat.

>  this might be a bad idea for fastcommits

Yes I was going to bring this up w.r.t. FastCommit. Our RecoverReadOnly in this case just needs to keep reading forward until the next commit record, backing up if corrupt record or no commit record found, instead of relying on commit.dat.

Reply:
> The dumb and easy fix in this case seems to be have readers retry on exceptions thrown -- purged commits will always be replaced by a later commit.

This might actually never happen (loop forever), because commits and deletions are happening so fast that the reader retries may never successfully capture a commit file. Also, note that the exception happens in the writer (unable to delete -- when we try to delete a file that the reader has open).

Reply:
We can probably punt on this problem for this PR. I personally think this should be part of the larger producer/consumer overhaul (e.g., with explicit Complete() calls for read iterators).

In this model, readonly fasterlog is just an iterator (or a bunch) on a difference process. Whatever fix we do that should apply here too with some generalization.

Reply:
Sounds good, we would likely want this overhaul to be in v2 as it might be a breaking change.